### ISSUE TITLE
XML not formed as expected

### ISSUE THREAD

**Issue by NuSkooler** [strongloop/strong-soap#141]
I'm calling a SOAP method and am getting an unexpected format sent out.

`client.Method(args, (err, result) => { ... } );` with args of:

```js
const args = {
  _this     : propertyCollector,
  specSet   : {
    propSet : {
      type    : 'VirtualMachine',
      pathSet : [ 'config' ]
    },
    objectSet : {
      obj : {
        $attributes : { type : 'ContainerView' },
        $value      : 'REDACTED',
      },
      skip            : true,
      selectSet : [
        {
          $attributes : { type : 'TraversalSpec', $xsiType : { xmlns : 'urn:vim25' } },
          $value      : {
            name    : 'traverseEntities',
            type    : 'ContainerView',
            path    : 'view',
            skip    : false,
          }
        }
      ]
    },
  },
  options : { }
};
```

Sends XML as follows:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Header/>
  <soap:Body>
    <ns1:RetrievePropertiesEx xmlns:ns1="urn:vim25">
      <ns1:_this type="PropertyCollector">propertyCollector</ns1:_this>
      <ns1:specSet>
        <ns1:propSet>
          <ns1:type>VirtualMachine</ns1:type>
          <ns1:pathSet>config</ns1:pathSet>
        </ns1:propSet>
        <ns1:objectSet>
          <ns1:obj type="ContainerView">REDACTED</ns1:obj>
          <ns1:skip>true</ns1:skip>
          <ns1:selectSet type="TraversalSpec">
            <ns1:name>traverseEntities</ns1:name>
            <type>ContainerView</type>
            <path>view</path>
            <skip>false</skip>
          </ns1:selectSet>
        </ns1:objectSet>
      </ns1:specSet>
      <ns1:options/>
    </ns1:RetrievePropertiesEx>
  </soap:Body>
</soap:Envelope>
```
(note missing `ns1` prefix on `type`, `path`, and `skip` above)

Expected XML:
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Header/>
  <soap:Body>
    <ns1:RetrievePropertiesEx xmlns:ns1="urn:vim25">
      <ns1:_this type="PropertyCollector">propertyCollector</ns1:_this>
      <ns1:specSet>
        <ns1:propSet>
          <ns1:type>VirtualMachine</ns1:type>
          <ns1:pathSet>config</ns1:pathSet>
        </ns1:propSet>
        <ns1:objectSet>
          <ns1:obj type="ContainerView">REDACTED</ns1:obj>
          <ns1:skip>true</ns1:skip>
          <ns1:selectSet type="TraversalSpec">
            <ns1:name>traverseEntities</ns1:name>
            <ns1:type>ContainerView</type>
            <ns1:path>view</path>
            <ns1:skip>false</skip>
          </ns1:selectSet>
        </ns1:objectSet>
      </ns1:specSet>
      <ns1:options/>
    </ns1:RetrievePropertiesEx>
  </soap:Body>
</soap:Envelope>
```

Any ideas? I've tried all sorts of approaches without any luck.

**Comment by raymondfeng** [comment]
@NuSkooler Please provide your wsdl/xsd. The following hints a unqualified local elements:
```xml
          <ns1:selectSet type="TraversalSpec">
            <ns1:name>traverseEntities</ns1:name>
            <type>ContainerView</type>
            <path>view</path>
            <skip>false</skip>
          </ns1:selectSet>
```

**Comment by NuSkooler** [comment]
@raymondfeng I believe this is what you're after: https://pastebin.com/UDKyCvTH

If I need to provide more, I'm happy to do so. VMWare's vCenter SOAP WSDL/XSD is split up into many query/files.

**Comment by raymondfeng** [comment]
Please provide complexType for `vim25:DynamicData`, which is missing from the xsd.

**Comment by NuSkooler** [comment]
@raymondfeng Thanks for the fast turn-around! Here you go: https://pastebin.com/NfjmdEcf

**Comment by raymondfeng** [comment]
Hmm, it seems that json2xml does not honor `xsi:type` to map subtype data. In your case, it is `TraversalSpec`, which extends from `SelectionSpec`.
```
<complexType name="TraversalSpec">
      <complexContent>
         <extension base="vim25:SelectionSpec">
            <sequence>
               <element name="type" type="xsd:string" />
               <element name="path" type="xsd:string" />
               <element name="skip" type="xsd:boolean" minOccurs="0" />
               <element name="selectSet" type="vim25:SelectionSpec" minOccurs="0" maxOccurs="unbounded" />
            </sequence>
         </extension>
      </complexContent>
   </complexType>
```

**Comment by NuSkooler** [comment]
@raymondfeng Any ideas on a work around? I haven't been able to get `$xml` to work at all for example, or I could just hack around it for now.

**Comment by raymondfeng** [comment]
We need to fix the code to take `xsi:type` into consideration to map sub types.

**Comment by NuSkooler** [comment]
@raymondfeng I'm happy to take a stab at whatever needs done, but I'll need some guidance in what I'm after :grin:

**Comment by raymondfeng** [comment]
It might be a bit involved. I'll do some investigation 1st. I'll let you know if/when you can help.

**Comment by raymondfeng** [comment]
@NuSkooler Please try out https://github.com/strongloop/strong-soap/pull/142

**Comment by NuSkooler** [comment]
@raymondfeng Looks like that did the trick!

**Comment by raymondfeng** [comment]
Fix released in `strong-soap@1.4.2`.

**Comment by NuSkooler** [comment]
@raymondfeng Thank you very much again for looking into this & getting it fixed quickly. This one change allows a project I'm working on to make a single SOAP call where I had to do hundreds recursively.

:1st_place_medal:
